ocaml/typing/subst.ml

```diff
@@ -188,118 +188,123 @@ let norm = function
 let ctype_apply_env_empty = ref (fun _ -> assert false)
 
 (* Similar to [Ctype.nondep_type_rec]. *)
-let rec typexp copy_scope s ty =
-  let desc = get_desc ty in
-  match desc with
-    Tvar _ | Tunivar _ ->
-      if s.for_saving || get_id ty < 0 then
-        let ty' =
-          if s.for_saving then newpersty (norm desc)
-          else newty2 ~level:(get_level ty) desc
-        in
-        For_copy.redirect_desc copy_scope ty (Tsubst (ty', None));
-        ty'
-      else ty
-  | Tsubst (ty, _) ->
-      ty
-  | Tfield (m, k, _t1, _t2) when not s.for_saving && m = dummy_method
-      && field_kind_repr k <> Fabsent && get_level ty < generic_level ->
-      (* do not copy the type of self when it is not generalized *)
-      ty
-(* cannot do it, since it would omit substitution
-  | Tvariant row when not (static_row row) ->
-      ty
-*)
-  | _ ->
-    let tm = row_of_type ty in
-    let has_fixed_row =
-      not (is_Tconstr ty) && is_constr_row ~allow_ident:false tm in
-    (* Make a stub *)
-    let layout = Layout.any in
-    let ty' =
-      if s.for_saving then newpersty (Tvar {name = None; layout})
-      else newgenstub ~scope:(get_scope ty) layout
-    in
-    For_copy.redirect_desc copy_scope ty (Tsubst (ty', None));
-    let desc =
-      if has_fixed_row then
-        match get_desc tm with (* PR#7348 *)
-          Tconstr (Pdot(m,i), tl, _abbrev) ->
-            let i' = String.sub i 0 (String.length i - 4) in
-            Tconstr(type_path s (Pdot(m,i')), tl, ref Mnil)
-        | _ -> assert false
-      else match desc with
-      | Tconstr (p, args, _abbrev) ->
-         let args = List.map (typexp copy_scope s) args in
-         begin match Path.Map.find p s.types with
-         | exception Not_found -> Tconstr(type_path s p, args, ref Mnil)
-         | Path _ -> Tconstr(type_path s p, args, ref Mnil)
-         | Type_function { params; body } ->
-            Tlink (!ctype_apply_env_empty params body args)
-         end
-      | Tpackage(p, fl) ->
-          Tpackage(modtype_path s p,
-                    List.map (fun (n, ty) -> (n, typexp copy_scope s ty)) fl)
-      | Tobject (t1, name) ->
-          let t1' = typexp copy_scope s t1 in
-          let name' =
-            match !name with
-            | None -> None
-            | Some (p, tl) ->
-                if to_subst_by_type_function s p
-                then None
-                else Some (type_path s p, List.map (typexp copy_scope s) tl)
+let typexp copy_scope s =
```

Comment:
For future reviewers: this whole change is much more easily reviewable with `patdiff`, it's just a change from `typeexp copy_scope s` to `loop` everywhere below